jib-core/src/main/java/com/google/cloud/tools/jib/registry/RegistryClient.java

```diff
@@ -242,46 +252,113 @@ public static Factory factory(
   }
 
   private final EventHandlers eventHandlers;
-  @Nullable private final Authorization authorization;
+  @Nullable private final Credential credential;
   private final RegistryEndpointRequestProperties registryEndpointRequestProperties;
   private final String userAgent;
   private final FailoverHttpClient httpClient;
 
+  // mutable
+  private final AtomicReference<Authorization> authorization = new AtomicReference<>();
+  private boolean readOnlyBearerAuth;
+
   /**
    * Instantiate with {@link #factory}.
    *
    * @param eventHandlers the event handlers used for dispatching log events
-   * @param authorization the {@link Authorization} to access the registry/repository
+   * @param credential credential for registry/repository; will not be used unless {@link
+   *     #configureBasicAuth} or {@link #doBearerAuth} is called
    * @param registryEndpointRequestProperties properties of registry endpoint requests
    * @param userAgent {@code User-Agent} header to send with the request
    * @param httpClient HTTP client
    */
   private RegistryClient(
       EventHandlers eventHandlers,
-      @Nullable Authorization authorization,
+      @Nullable Credential credential,
       RegistryEndpointRequestProperties registryEndpointRequestProperties,
       String userAgent,
       FailoverHttpClient httpClient) {
     this.eventHandlers = eventHandlers;
-    this.authorization = authorization;
+    this.credential = credential;
     this.registryEndpointRequestProperties = registryEndpointRequestProperties;
     this.userAgent = userAgent;
     this.httpClient = httpClient;
   }
 
+  public void configureBasicAuth() {
+    Preconditions.checkNotNull(credential);
+    Preconditions.checkState(!credential.isOAuth2RefreshToken());
+
+    authorization.set(
+        Authorization.fromBasicCredentials(credential.getUsername(), credential.getPassword()));
+
+    String registry = registryEndpointRequestProperties.getServerUrl();
+    String repository = registryEndpointRequestProperties.getImageName();
+    eventHandlers.dispatch(
+        LogEvent.debug("configured basic auth for " + registry + "/" + repository));
+  }
+
   /**
-   * @return the {@link RegistryAuthenticator} to authenticate pulls/pushes with the registry, or
-   *     {@link Optional#empty()} if no token authentication is necessary
+   * Attempts bearer authentication.
+   *
+   * @param readOnlyBearerAuth true to request pull scope only; false for pull and push
```

Comment:
should readOnly be a fixed value of a registry client? Do we intend to reuse a registry client to do readonly once and then write later?
Perhaps this can go on the factory?

Reply:
As the name suggested, this read-only vs. read-write applies only when doing bearer auth. Before doing any auth or in the case of basic auth, the registry client itself is unrestricted. It will then be the server who will say you cannot do anything, can read, or can read and write.

The class is mutable, and potentially it's possible to do `doBearerAuth(true)` and later `doBearerAuth(false)`, but I never intend to do it. But the point is, the client doesn't know if it's going to be read only (actually just requesting read only) before doing `doBearerAuth()`. And it is a bit sad, but I don't think the factory should initiate `doBearerAuth()`.